gapis/trace/android/trace.go

```diff
@@ -50,12 +71,137 @@ type androidTracer struct {
+		return &adreno.AdrenoValidator{}
+	}
+	return nil
+}
+
+func traceOptions(ctx context.Context, v validate.Validator) (*service.TraceOptions, error) {
+	counters := v.GetCounters()
+	ids := make([]uint32, len(counters))
+	for i, counter := range counters {
+		ids[i] = counter.Id
+	}
+	return &service.TraceOptions{
+		DeferStart: true,
+		PerfettoConfig: &perfetto_pb.TraceConfig{
+			Buffers: []*perfetto_pb.TraceConfig_BufferConfig{
+				{SizeKb: proto.Uint32(bufferSizeKb)},
+			},
+			DurationMs: proto.Uint32(durationMs),
+			DataSources: []*perfetto_pb.TraceConfig_DataSource{
+				{
+					Config: &perfetto_pb.DataSourceConfig{
+						Name: proto.String(gpuRenderStagesDataSourceDescriptorName),
+					},
+				},
+				{
+					Config: &perfetto_pb.DataSourceConfig{
+						Name: proto.String(gpuCountersDataSourceDescriptorName),
+						GpuCounterConfig: &perfetto_pb.GpuCounterConfig{
+							CounterPeriodNs: proto.Uint64(counterPeriodNs),
+							CounterIds:      ids,
+						},
+					},
+				},
+			},
+		},
+	}, nil
 }
 
 func (t *androidTracer) GetDevice() bind.Device {
 	return t.b
 }
 
+func (t *androidTracer) Validate(ctx context.Context) error {
+	ctx = status.Start(ctx, "Android Device Validation")
+	defer status.Finish(ctx)
+	if t.v == nil {
+		return log.Errf(ctx, nil, "No validator found for device %d", t.b.Instance().ID.ID())
+	}
+	d := t.b.(adb.Device)
+	osConfiguration := d.Instance().GetConfiguration()
+	if osConfiguration.GetPerfettoCapability() == nil {
+		return log.Errf(ctx, nil, "No Perfetto Capability found on device %d", d.Instance().ID.ID())
+	}
+	if gpuProfiling := osConfiguration.GetPerfettoCapability().GetGpuProfiling(); gpuProfiling == nil || gpuProfiling.GetGpuCounterDescriptor() == nil {
+		return log.Errf(ctx, nil, "No GPU profiling capabilities found on device %d", d.Instance().ID.ID())
+	}
+
+	// Get ActivityAction
+	packages, _ := d.InstalledPackages(ctx)
+	pkg := packages.FindByName(validationPackageName)
+	if pkg == nil {
+		return log.Errf(ctx, nil, "Package %v not found", validationPackageName)
+	}
+	activityAction := pkg.ActivityActions.FindByName(intentAction, activityName)
+	if activityAction == nil {
+		return log.Errf(ctx, nil, "Activity %v not found in %v", activityName, validationPackageName)
+	}
+
+	// Construct trace config
+	traceOpts, err := traceOptions(ctx, t.v)
+	if err != nil {
+		return log.Err(ctx, err, "Could not get the trace configuration")
+	}
+	process, cleanup, err := perfetto_android.Start(ctx, d, activityAction, traceOpts)
+	if err != nil {
+		cleanup.Invoke(ctx)
+		return log.Err(ctx, err, "Error when start Perfetto tracing.")
+	}
+	defer cleanup.Invoke(ctx)
+
+	// Force to stop the application, ignore any error that happens as it
+	// doesn't affect validation.
+	defer d.ForceStop(ctx, validationPackageName)
+
+	var buf bytes.Buffer
+	var written int64
+	// Start to capture.
+	status.Do(ctx, "Android Device Validation: Tracing", func(ctx context.Context) {
```

Comment:
Don't repeat the name. Just call it `Tracing`. The status package creates a tree, and `Tracing` will be a child of `Android Device Validation`.

ditto below